<html>
<head>
	<link rel="stylesheet" href="https://cdn.jsdelivr.net/npm/@picocss/pico@2/css/pico.min.css"/>
	<title>Lecture "Applied Mathematics in Biology: Single-cell omics"</title>
</head>
<body>
<main class="container">

<header>
<p style="text-align:right">
	<a href="https://www.bioquant.uni-heidelberg.de"><img src="https://www.bioquant.uni-heidelberg.de/themes/custom/bootstrap_bq/logo.svg" style="width:20%;padding:3%"></a>
	<a href="https://uni-heidelberg.de"><img src="https://www.bioquant.uni-heidelberg.de/themes/custom/bootstrap_bq/logo_hd.svg" style="width:20%;padding:3%"></a></p>

</header>

<h4>Lecture Announcement:</h4>
<h2 style="text-align:center;padding:1%">Applied Mathematics in Biology: Single-cell omics</h2>
<i>Lecture, summer term 2024</i>


<p><b>Lecturer: </b><a href="https://www.zmbh.uni-heidelberg.de/Anders/">Simon Anders</a>, <a href="https://www.bioquant.uni-heidelberg.de/">BioQuant</a>, <a href="mailto:simon.anders@bioquant.uni-heidelberg.de">simon.anders@bioquant.uni-heidelberg.de</a></p>

<p><b>Target audience:</b> students of math-oriented subjects (mathematics, physics, computer science, engineering, scientific computing); also life-sciences students with advanced math knowledge</p>

<p><i><b>This lecture is aimed at students of computational sciences interested in applying their skills in the life sciences. It is designed to demonstrate such interdisplinary work on an example topic and use this to dive into various areas of applied mathematics and of biology.</b></i></p>

<h4>Topic and Aims</h4>

<p>Many experimental techniques of modern biology need sophisticated data analysis methods. Developing these requires, among others, mastery of methods of mathematical modelling, of stochastic models, of numerics and scientific software engineering. Therefore, such methodolical research falls outside the expertise of most biologists; and is a task for applied mathematicians, physicists or math-oriented computer scientists.</p>

<p>One particular topic of currently rapidly growing importance is "single-cell omics": For several tissue samples, each comprising hundreds to thousands of cells, one measures for each cell a large vector of values that represent, e.g., the activity of genes or the epigenetic state in cells. This allows to, e.g., compare, healthy and diseased tissues, understand the make-up of tumours, track the embryonic development of organs, etc. In such data, each cell is represented by a vector in a feature space (typically simply ) and the aim is to explore the geometry and topology of the manifolds sampled by the cells and translate the findings into the language of biology, and to perform statistical inference to gain generalizable conclusions about the experiment.</p>

<p>An aim of the lecture is therefore to introduce students to this area of work for researchers in applied mathematics. However, rather than giving a general overview, we will explore a specific field, namely single-cell omics, in depth, and learn not only how to bring modern methods of mathematics and scientific computing to bear on a problems of practical importance but also how to grapple with the communication difficulties of building bridges between two disciplines, mathematics and biology, with different language, concepts and ways of thinking.</p>

<p>Starting with the basics of the relevant parts of biology and applied math, we will advance to the current methodological frontier of the field.</p>

<h4>Schedule und Registration</h4>

<p><s>There will be a weekly lecture of 90 minutes and a problem/exercise class, also 90 minutes. Times and dates will be determined by online poll. If you are interested, please send <a href="mailto:simon.anders@bioquant.uni-heidelberg.de">me</a> an e-mail so I can include you in the poll.</s></p>

<p>If you intend to participate, please join the course on <a href="https://moodle.uni-heidelberg.de/course/view.php?id=21765">Moodle</a>

<p>Lectures will be every Monday, 4pm to 6pm, in seminar room 042 in BioQuant (INF 267). For the exercise class, we meet at the same room on Wednesdays at 10am-12. We start on 22 April.</p>

<h4>Credits</h4>

<ul>
	<li>for students in Master programmes <i>mathematics</i> and <i>scientific computing</i>: tbd (6 LP ?)</li>
	<li>for students of physics: see rules for math as minor subject</li>
	<li>for students of biology and others: no rules established yet; please ask me or your study coordinator</li>
</ul>

<p>There will be a written or oral exam at the end of term; with the possible alternative of a homework project.</p>

<h4>Prerequisites</h4>

<p>Recommended is a basic knowledge in (finite-dimensional) linear algebra
(matrix calculus, eigendecomposition, etc), experience with at least one
programming language and interest in biology</p>

<h4>Contents</h4>
<p>Content of this lecture is:</p>
<ul>
<li>basics of molecular biology, as needed</li>
<li>assay techniques in single cell (sc) biology</li>
<li>stochastic models for omics data</li>
<li>the concept of feature space, as used for omics data and in machine
learning</li>
<li>methods to explore high-dimensional data</li>
<li>linear and non-linear methods for dimension reductions</li>
<li>clustering in high-dimensional space</li>
<li>non-linear regression methods</li>
<li>graph-based methods in omics data analysis</li>
<li>applications of machine learning methods to sc omics</li>
<li>deep learning (esp. variational autoencoders) and related methods</li>
<li>technqiues for handling big data</li>
<li>interactive visualization for exploratory data analysis</li>
</ul>

<h4>Learning goals</h4>

<p>
Modern molecular biology has much need for computational expertise, providing opportunities for researchers from mathematics, physics, scientific computing or other computational sciences. This lecture aims to provide knowledge and skills to get started with such interdisciplinary work &mdash; by diving into one specific topic, namely the analysis of single-cell omics data, i.e., mRNA sequencing data with single-cell resolution, a technique used to characterize the molecular state of cells in tissue samples, in order to understand biological processes in health and disease.</p>

<p>Students will learn how to work in a foreign scientific domain of knowledge (biology), apply mathematical concepts there, bridge terminology gaps and other overcome challenges of interdisciplinary work. We will explore "real-world" applications of mathematical methods, especially from high-dimensional statistics, and learn how to use exploratory data analysis to turn domain-language (here: biological) questions into mathematical hypotheses on the structure of the given data manifolds, and how to translate results back to make them understandable for domain experts</p>

<p>Students will gain practical experience in scientific programming, applied linear algebra, machine learning, data reduction and interpretation, and, most importantly, in interdisciplinary research work, as well as insights into opportunities for computational scientists to work in the life sciences.</p>

</main>
<footer><p style="text-align:right;padding-right:5%">Last change: 2024-03-26</footer>
</body></html>
